pyro/util.py

```diff
@@ -346,6 +349,8 @@ def check_model_guide_match(model_trace, guide_trace):
 
 
 def check_site_shape(site, max_iarange_nesting):
+    if not pyro.infer.is_validation_enabled():
```

Comment:
I think it's cleaner to move the gating to outside of the checking logic. E.g `check_site_shape` is called inside loops in `Trace_ELBO` ([code](https://github.com/uber/pyro/blob/dev/pyro/infer/trace_elbo.py#L46)), and those loops can be entirely avoided:
```py
if not pyro.infer.is_validation_enabled():
    for site in ...:
       if ...:
            check_site_shape(site, ...)
    for site in ...:
       if ...:
            check_site_shape(site, ...)
```

Reply:
if site shapes dont match, it will always result in an exception correct? if so, you need this check regardless of validation level

Reply:
I think the idea is to trade off computation for clarity of error messages: if `_VALIDATION_ENABLED` then perform possibly-expensive checks to ensure errors are caught as early as possible. Note that some shape errors might result in silent failure due to broadcasting.